tchannel-core/src/main/java/com/uber/tchannel/codecs/CodecUtils.java

```diff
@@ -150,4 +154,144 @@ public static void encodeTrace(Trace trace, ByteBuf buffer) {
+
+        return comp;
+    }
+
+    public static ByteBuf writeArgCopy(ByteBufAllocator allocator, ByteBuf payload, ByteBuf arg, int writableBytes) {
+        if (writableBytes <= 2) {
+            throw new UnsupportedOperationException("writableBytes must be larger than 2");
+        }
+
+        int readableBytes = arg.readableBytes();
+        int headerSize = 2;
+        int chunkLength = Math.min(readableBytes + headerSize, writableBytes);
+
+        // Write the size of the `arg`
+        payload.writeShort(chunkLength - headerSize);
+        if (readableBytes == 0) {
+            return payload;
+        } else {
+            return payload.writeBytes(arg, chunkLength - headerSize);
+        }
+    }
+
+    public static ByteBuf writeArgsCopy(ByteBufAllocator allocator,
+                                    ByteBuf header,
+                                    List<ByteBuf> args) {
+        ByteBuf payload = allocator.buffer(header.readableBytes(), TFrame.MAX_FRAME_PAYLOAD_LENGTH);
+        payload.writeBytes(header);
+        header.release();
+        int writableBytes = TFrame.MAX_FRAME_PAYLOAD_LENGTH - payload.readableBytes();
+
+        while (!args.isEmpty()) {
+            ByteBuf arg = args.get(0);
+            writeArgCopy(allocator, payload, arg, writableBytes);
+            writableBytes = TFrame.MAX_FRAME_PAYLOAD_LENGTH - payload.readableBytes();
+            if (writableBytes <= 2) {
+                break;
+            }
+
+            if (arg.readableBytes() == 0) {
+                args.remove(0);
+            }
+        }
+
+        return payload;
+    }
+
+    public static ByteBuf compose(ByteBuf first, ByteBuf second) {
+        if (first == Unpooled.EMPTY_BUFFER) {
+            return second;
+        } else if (second == Unpooled.EMPTY_BUFFER) {
+            return first;
+        } else {
+            return Unpooled.wrappedBuffer(first, second);
+        }
+    }
+
+    public static ByteBuf readArg(ByteBuf buffer) {
+        if (buffer.readableBytes() < 2) {
+            return null;
+        }
+
+        int len = buffer.readUnsignedShort();
+        if (len > buffer.readableBytes()) {
+            throw new UnsupportedOperationException("wrong read index for args");
+        } else if (len == 0) {
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        /* Read a slice, retain a copy */
+        ByteBuf arg = buffer.readSlice(len);
+        arg.retain();
+
+        return arg;
+    }
+
+    public static void readArgs(List<ByteBuf> args, ByteBuf buffer) {
+
+        ByteBuf arg = null;
+        if (args.isEmpty()) {
+            args.add(Unpooled.EMPTY_BUFFER);
+        }
+
+        boolean first = true;
+        while (true) {
+            arg = readArg(buffer);
+            if (arg == null) {
+                return;
+            } else if (first) {
+                first = false;
+                ByteBuf prev = args.get(args.size() - 1);
```

Comment:
actually, it cannot be a queue, since we want random access to it, e.g., get arg2.